docs/Tutorial_DAMP.ipynb

```diff
@@ -0,0 +1,592 @@
+{
```

Comment:
<pre class="ql-syntax" spellcheck="false">Line #4.        chunks, each with a length of power of two. Each chunk is
&nbsp; &nbsp; preceded by one that is twice in size
</pre><p>So, for <code>naive_get_range_damp(stop=100, m=5)</code> what should the output be? I am seeing:</p><pre class="ql-syntax" spellcheck="false">array([[ 92, 100],
&nbsp; &nbsp; &nbsp; &nbsp;[ 81,&nbsp; 97],
&nbsp; &nbsp; &nbsp; &nbsp;[ 54,&nbsp; 86],
&nbsp; &nbsp; &nbsp; &nbsp;[&nbsp; 0,&nbsp; 59]])
</pre><p>but the difference in chunk size is:</p><pre class="ql-syntax" spellcheck="false">array([ 8, 16, 32, 59])
</pre><p>Which is inconsistent with your description. Why is the final one <code>59</code> and then why is each one preceded by HALF in size? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/920/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Reply:
@seanlaw 
We have two things here:
(1) The direction we traverse the time series data (which is backward here, i.e. from `stop` to `start=0`)
(2) The order we represent the chunks range in the numpy array

> Which is inconsistent with your description
> why is each one preceded by HALF in size?

I think (1) and (2) got mixed. To better explain this, let's reverse the order of the chunks returned by the function `naive_get_range_damp`:

```
out = naive_get_range_damp(stop=100, m=5)
out = out[::-1]
```

which gives:
```
>>> out
array([
       [  0,  59], 
       [ 54,  86],
       [ 81,  97],
       [ 92, 100]
])
```

Now, we can see that as we move in the time series from right (`stop=100`) to left (`start=0`), AND starts from the last row of `out`, then each chunk is PRECEDED BY TWICE in size (except the first one!)

Note that the length of the chunk `out[1]` (with the new order above) is 32. So, the most-left chunk, i.e. `out[0]`, should have a length of 64. But, it is just truncated at index 0.

Reply:
# The diff between stop-start should always be an incrementing power of 2

The problem is how do we know the set of "incrementing powers of 2" that will fit into the `range(start, stop)`? First, we must make the simpler observation that, in the worst case scenario, the cumulative sum of the "incrementing powers of 2" must be (just) larger than or equal to `stop-start`. For example, if `start=0` and `stop=100`, then we need to look for some set of `1 + 2 + 4 + ... + n` that will be just larger than or equal to `stop - start = 100` so that it can act as a bound for our `for` loop. However, after staring at the cumulative sum for different values `n`, I noticed a strange pattern:

| n | cumsum | pattern |
|---|--------|---------|
| 0 | 0      | $2^0-1$ |
| 1 | 1      | $2^1-1$ |
| 2 | 3      | $2^2-1$ |
| 3 | 7      | $2^3-1$ |
| 4 | 15     | $2^4-1$ |
| 5 | 31     | $2^5-1$ |
| 6 | 63     | $2^6-1$ |

And so, for any `n`, the pattern for the cumulative sum of the "incrementing powers of 2" is simply $2^n-1$! Again, we must solve for $n_{upper}$ (upper bound) under the condition that the cumulative sum is (just) larger than or equal to `stop - start = 100`:

$$
\begin{align}
2^{n_{upper}}-1 &\ge 100
\newline 2^{n_{upper}} &\ge 101
\newline \log_22^{n_{upper}} &\ge log_2101
\newline {n_{upper}} &\ge log_2101
\newline {n_{upper}} &= \lceil{log_2101}\rceil
\newline {n_{upper}} &= \lceil{6.658211482751795}\rceil
\newline {n_{upper}} &= 7
\end{align}
$$

# Example

```
start, stop = 0, 100
n_upper = math.ceil(math.log2(stop-start+1))
n_upper

# 7
```

Of course, you may have noticed that we should NOT have started our cumulative sum at `1` and, instead, we should start our cumulative sum at the next power of 2 that is closest to the window size, `m`. So, if `m=5`, then we should start the cumulative sum at $n_{lower}=3$. This is easily remedied by subtracting the cumulative sum leading up to `8`:

$$
\begin{align}
\left( 2^{n_{upper}}-1 \right) - \left( 2^{n_{lower}}-1 \right) &\ge 100
\newline  2^{n_{upper}}-1 &\ge 100 + \left( 2^{n_{lower}}-1 \right)
\newline  2^{n_{upper}}-1 &\ge 100 + \left( 2^3-1 \right)
\newline  2^{n_{upper}}-1 &\ge 100 + \left( 8-1 \right)
\newline  2^{n_{upper}}-1 &\ge 100 + 7
\newline  2^{n_{upper}} &\ge 107 + 1
\newline  \log_22^{n_{upper}} &\ge log_2108
\newline  {n_{upper}} &\ge log_2108
\newline  {n_{upper}} &= \lceil{log_2108}\rceil
\newline  {n_{upper}} &= \lceil{6.754887502163468}\rceil
\newline  {n_{upper}} &= 7
\end{align}
$$

```
m = 5
chunk_start = next_pow2(m)
n_lower = math.ceil(math.log2(chunk_start))
n_upper = math.ceil(math.log2(stop + 2**n_lower - 1 + 1))
n_upper

# 7
```

Applying this, we get a new function:

```
def get_damp_range(stop, m, start=0):
    out = []
    n_lower = math.ceil(math.log2(next_pow2(m)))
    n_upper = math.ceil(math.log2(stop + 2**n_lower)) + 1
    chunk_stop = stop
    
    for chunk_size in np.power(2, range(n_lower, n_upper)):
        chunk_start = chunk_stop - chunk_size
        out.append([max(chunk_start, start), chunk_stop])
        chunk_stop = chunk_start + m

        if chunk_start <= 0:
                break
        
    return np.array(out)
```

And we can verify that the results are the same:

```
import numpy.testing as npt

for stop in range(1, 1000):
    for m in range(1, 25):
        ref = naive_get_range_damp(stop, m)
        cmp = get_damp_range(stop, m)
        npt.assert_almost_equal(ref, cmp)
```

All of the tests pass. Hopefully, I didn't make too many mistakes.

Reply:
**Note 1:**
I noticed that two adjacent chunks range have one subsequence in common.
```
# stop=100, 
# m =5
>>> out
array([[ 92, 100],
       [ 81,  97],
       [ 54,  86],
       [  0,  59]])
```

The subsequence `T[92:97]` is in both chunks `[92, 100]` and `[81, 97]`.  This can be resolved by changing the line `chunk_stop = chunk_start + m` to `chunk_stop = chunk_start + m - 1`


**Note 2:**
@seanlaw 
I think I get what you did. But, still... there is something that I cannot get my head around. 

> in the worst case scenario, the cumulative sum of the "incrementing powers of 2" must be (just) larger than or equal to stop-start

I am not sure about this. Let's say `start` is set to `0`, and $n_{lower}$ is the minimum integer value that satisfies $2^{n_{lower}} \ge m$. Note that the total number of subsequences is: `stop - m + 1`. The size of each chunk is power of two, i.e. $2^{num}$. Hence, the NUMBER of SUBSEQUENCES within this chunk is: $2^{num} - m + 1$. So, we have:

$(2^{n_{lower}}  - m + 1) +  (2 * 2^{n_{lower}}  - m + 1) + ... + (2^{n-1} * 2^{n_{lower}} - m + 1) \ge stop -m + 1$
where `n` is the number of chunks. The left side is the cumulative sum of number of subsequences in chunks, and the right side is the total number of subsequences.

Hence:
$2^{n_{lower}} \left(1 + 2 + 4 + ... + 2^{n-1} \right) - m * n +  n   \ge stop - m + 1$
$2^{n_{lower}} (2^{n} - 1) - mn + n \ge stop - m + 1$
$2^{n_{lower}} (2^{n} - 1)  \ge stop + (n - 1)(m - 1)$


We can convert the inequation above to the following format: $Ae^{x} + Bx + C = 0$, and solve it. I do not want to go to details. I just wanted to mention that this is different than what you obtained, which is: 
$2^{n_{lower}} (2^{n} - 1)  \ge stop$
$2^{n_{lower} + n}  \ge stop + 2 ^ {n_{lower}}$
where, $n_{lower} + n$ is $n_{upper}$.

Reply:
> I'm not sure what the problem is. Both naive_get_range_damp and naive_get_range_damp produce identical results. So, I'm just following what you are algorithm is doing. Feel free to adjust accordingly.

That was a general note (like a self note so that I do not forget).  I think I should have mentioned in a separate comment section to avoid creating confusion.

 
> So, my inequality just an upper bound for the for loop

Yes...I think I was trying to confirm that. I think I can now see what you did!...
$2^{n_{lower}} (2^{n} - 1)  \ge 2^{n_{lower}} (2^{n} - 1)  - (n - 1)(m - 1) \ge stop $

In fact, when you said " upper bound", it helps me to understand what you meant by 
> First, we must make the simpler observation that, in the worst case scenario, the cumulative sum of the "incrementing powers of 2" must be (just) larger than or equal to stop-start

Instead of $2^{num} - m + 1$, you just considered the upper bound $2 ^ {num}$! 

> Maybe you disagree but I think my proposal is pretty straightforward to comprehend.

I think the lock is unlocked in my brain 😄 I think your solution (with that upper bound trick) is waaaay more cleaner.

Reply:
Right. I haven't thought this through but since we are constantly shifting the stop index by `m`, the upper bound might be even worse if `len(T)` is super long and `m` is super small since there will be a ton of overlap between ranges. Maybe we should test this (as I've only used a small `len(T)` and small `m`. Any issues may reveal itself when `len(T)` is large and `m` is small (or large?!?).